### DISCUSSION TITLE
Here is how to use UUIDs as directory names instead of regular IDs

### DISCUSSION THREAD

**Discussion by miclf** [spatie/laravel-medialibrary#1921] [Show and tell] [answered]
### tl;dr

[Make your own implementation](https://docs.spatie.be/laravel-medialibrary/v8/advanced-usage/using-a-custom-directory-structure/) of `PathGenerator` and use `$media->getAttribute('uuid')` instead of `$media->getKey()` to determine the base path.

---

Hi there,

I’ve just upgraded from `v7` to `v8` in a new project I’m working on and I thought maybe some people would be interested in this little trick.

You may know that version 8 automatically generates a UUID for each media. This is really useful if, for example, you want to prevent users to tweak URLs to discover (or to scrape) all of your images. But, [by default, MediaLibrary will put each media (and its conversions) in a directory named with the ID (primary key) of the media](https://docs.spatie.be/laravel-medialibrary/v8/advanced-usage/using-a-custom-directory-structure/).

If you want to use UUIDs instead, you have to add a bit of customization. Here is how to do it.

### How MediaLibrary decides where to store files

In order to determine where to store and retrieve your files, the package uses a class called [`DefaultPathGenerator`](https://github.com/spatie/laravel-medialibrary/blob/8.3.0/src/Support/PathGenerator/DefaultPathGenerator.php). It contains a few simple methods telling how to retrieve files for medias, conversions and responsive images.

In this file, we can see that the ID (the primary key) of each media model is used to name each media directory:

```php
/*
 * Get a unique base path for the given media.
 */
protected function getBasePath(Media $media): string
{
    return $media->getKey();
}
```

The `getKey()` is a method from Eloquent, which retrieves the value of the primary key. This is where we will need to make a change in order to use the `uuid` column instead.

But we cannot update that file directly, obviously, because it is in the `vendor/` directory, and you would loose all your changes as soon as you update the package using Composer. So we will create our own version of this file instead.

### Create a custom `PathGenerator`

Create a copy of the [`DefaultPathGenerator`](https://github.com/spatie/laravel-medialibrary/blob/8.3.0/src/Support/PathGenerator/DefaultPathGenerator.php) class somewhere in your `app/` directory. You can put it wherever you want. In this example, I’m putting it in a subdirectory: `app/MediaLibrary/`, and my class will be named `CustomPathGenerator`.

I then need to:
1. update the namespace to reflect where I created the file ;
2. import the `PathGenerator` interface to make the package know that my class provides the methods it needs ;
3. modify the class name if needed.

In this example, here is what I get:

```php
<?php

namespace App\MediaLibrary; // Step 1.

use Spatie\MediaLibrary\MediaCollections\Models\Media;
use Spatie\MediaLibrary\Support\PathGenerator\PathGenerator; // Step 2.

// Step 3.
class CustomPathGenerator implements PathGenerator
{
    // All the code inside my class...
}
```

### Tell the package to use the `uuid` column to name media directories

Now we can update the `getBasePath()` method as needed:

```php
/*
 * Get a unique base path for the given media.
 */
protected function getBasePath(Media $media): string
{
    return $media->getAttribute('uuid');
}
```

There is just one final thing we need to do: make MediaLibrary use our custom path generator instead of its default one.

### Update the package configuration

In your [`config/media-library.php`](https://github.com/spatie/laravel-medialibrary/blob/8.3.0/config/media-library.php) file, look for the `path_generator` setting.

If you don’t have this configuration file, you first need to [publish it from the package](https://docs.spatie.be/laravel-medialibrary/v8/installation-setup/):

```bash
php artisan vendor:publish --provider="Spatie\MediaLibrary\MediaLibraryServiceProvider" --tag="config"
```

Then, replace the value of the `path_generator` setting to make it use your own class. In this example:
```php
/*
 * The class that contains the strategy for determining a media file's path.
 */
'path_generator' => \App\MediaLibrary\CustomPathGenerator::class,
```

And voilà!

Obviously, this small tutorial doesn’t cover how to rename existing directories, etc. The ideal use case is to do that on new projects or in applications that don’t store media yet.

You can now further modify your custom class in case you want to customize where conversions or responsive images are stored.

**Comment by freekmurze** [comment, marked as answer]
Thanks!

**Comment by dominicbachmann** [comment]
Thanks! To make it work right now instead of implementing "PathGenerator" you should extend "DefaultPathGenerator".

**Reply by miclf** [reply]
You could indeed extend a built-in class.

The tutorial here prefers to *implement an interface* rather than *extend a class*, but both approaches are valid.

In theory, interfaces change much less often than classes, which usually makes them more appropriate when trying to replace of extend features as we do here.

**Comment by francoism90** [comment]
Do not forget to update the config as well:
```
/*
     * Moves media on updating to keep path consistent. Enable it only with a custom
     * PathGenerator that uses, for example, the media UUID.
     */
    'moves_media_on_update' => true,
```

Edit: Maybe this is not needed, only when use a custom path generator? But hopefully it's helpful to mention. :)

**Reply by usmanarain350** [reply]
Yeah it saves my  time Thanks Dude @francoism90

**Comment by vz-spr** [comment]
Thank you!

**Comment by modrink** [comment]
I'm using S3 - it seems that updating the directory structure configuration works OK, original file upload generates the structure correctly, getTemporaryUrl() method works and I can access the file, but when it comes to generating responsive images via a job, it fails because it tries to reach the original file via the media ID (not UUID).

Did I do something wrong?

```Exception
GuzzleHttp\Exception\ClientException: Client error: `GET https://s3.url.com/148/zHSTnZgnnXCnumAT.jpg` resulted in a `404 Not Found` response:
<?xml version="1.0" encoding="UTF-8"?><Error><Code>NoSuchKey</Code><Message>The specified key does not exist.</Message>< (truncated...)
in /var/www/html/vendor/guzzlehttp/guzzle/src/Exception/RequestException.php:113
Stack trace:
#0 /var/www/html/vendor/guzzlehttp/guzzle/src/Middleware.php(69): GuzzleHttp\Exception\RequestException::create()
#1 /var/www/html/vendor/guzzlehttp/promises/src/Promise.php(204): GuzzleHttp\Middleware::GuzzleHttp\{closure}()
#2 /var/www/html/vendor/guzzlehttp/promises/src/Promise.php(153): GuzzleHttp\Promise\Promise::callHandler()
#3 /var/www/html/vendor/guzzlehttp/promises/src/TaskQueue.php(48): GuzzleHttp\Promise\Promise::GuzzleHttp\Promise\{closure}()
#4 /var/www/html/vendor/guzzlehttp/guzzle/src/Handler/CurlMultiHandler.php(159): GuzzleHttp\Promise\TaskQueue->run()
#5 /var/www/html/vendor/guzzlehttp/guzzle/src/Handler/CurlMultiHandler.php(184): GuzzleHttp\Handler\CurlMultiHandler->tick()
#6 /var/www/html/vendor/guzzlehttp/promises/src/Promise.php(248): GuzzleHttp\Handler\CurlMultiHandler->execute()
#7 /var/www/html/vendor/guzzlehttp/promises/src/Promise.php(224): GuzzleHttp\Promise\Promise->invokeWaitFn()
#8 /var/www/html/vendor/guzzlehttp/promises/src/Promise.php(269): GuzzleHttp\Promise\Promise->waitIfPending()
#9 /var/www/html/vendor/guzzlehttp/promises/src/Promise.php(226): GuzzleHttp\Promise\Promise->invokeWaitList()
#10 /var/www/html/vendor/guzzlehttp/promises/src/Promise.php(269): GuzzleHttp\Promise\Promise->waitIfPending()
#11 /var/www/html/vendor/guzzlehttp/promises/src/Promise.php(226): GuzzleHttp\Promise\Promise->invokeWaitList()
#12 /var/www/html/vendor/guzzlehttp/promises/src/Promise.php(62): GuzzleHttp\Promise\Promise->waitIfPending()
#13 /var/www/html/vendor/aws/aws-sdk-php/src/AwsClientTrait.php(58): GuzzleHttp\Promise\Promise->wait()
#14 /var/www/html/vendor/league/flysystem-aws-s3-v3/AwsS3V3Adapter.php(499): Aws\AwsClient->execute()
#15 /var/www/html/vendor/league/flysystem-aws-s3-v3/AwsS3V3Adapter.php(262): League\Flysystem\AwsS3V3\AwsS3V3Adapter->readObject()
#16 /var/www/html/vendor/league/flysystem/src/Filesystem.php(79): League\Flysystem\AwsS3V3\AwsS3V3Adapter->readStream()
#17 /var/www/html/vendor/laravel/framework/src/Illuminate/Filesystem/FilesystemAdapter.php(588): League\Flysystem\Filesystem->readStream()
#18 /var/www/html/vendor/spatie/laravel-medialibrary/src/MediaCollections/Filesystem.php(180): Illuminate\Filesystem\FilesystemAdapter->readStream()
#19 /var/www/html/vendor/spatie/laravel-medialibrary/src/MediaCollections/Filesystem.php(185): Spatie\MediaLibrary\MediaCollections\Filesystem->getStream()
#20 /var/www/html/vendor/spatie/laravel-medialibrary/src/Conversions/FileManipulator.php(57): Spatie\MediaLibrary\MediaCollections\Filesystem->copyFromMediaLibrary()
#21 /var/www/html/vendor/spatie/laravel-medialibrary/src/Conversions/Jobs/PerformConversionsJob.php(27): Spatie\MediaLibrary\Conversions\FileManipulator->performConversions()
#22 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(36): Spatie\MediaLibrary\Conversions\Jobs\PerformConversionsJob->handle()
#23 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/Util.php(41): Illuminate\Container\BoundMethod::Illuminate\Container\{closure}()
#24 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(93): Illuminate\Container\Util::unwrapIfClosure()
#25 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(37): Illuminate\Container\BoundMethod::callBoundMethod()
#26 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/Container.php(651): Illuminate\Container\BoundMethod::call()
#27 /var/www/html/vendor/laravel/framework/src/Illuminate/Bus/Dispatcher.php(128): Illuminate\Container\Container->call()
#28 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(141): Illuminate\Bus\Dispatcher->Illuminate\Bus\{closure}()
#29 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(116): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}()
#30 /var/www/html/vendor/laravel/framework/src/Illuminate/Bus/Dispatcher.php(132): Illuminate\Pipeline\Pipeline->then()
#31 /var/www/html/vendor/laravel/framework/src/Illuminate/Queue/CallQueuedHandler.php(124): Illuminate\Bus\Dispatcher->dispatchNow()
#32 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(141): Illuminate\Queue\CallQueuedHandler->Illuminate\Queue\{closure}()
#33 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(116): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}()
#34 /var/www/html/vendor/laravel/framework/src/Illuminate/Queue/CallQueuedHandler.php(126): Illuminate\Pipeline\Pipeline->then()
#35 /var/www/html/vendor/laravel/framework/src/Illuminate/Queue/CallQueuedHandler.php(70): Illuminate\Queue\CallQueuedHandler->dispatchThroughMiddleware()
#36 /var/www/html/vendor/laravel/framework/src/Illuminate/Queue/Jobs/Job.php(98): Illuminate\Queue\CallQueuedHandler->call()
#37 /var/www/html/vendor/laravel/framework/src/Illuminate/Queue/Worker.php(425): Illuminate\Queue\Jobs\Job->fire()
#38 /var/www/html/vendor/laravel/framework/src/Illuminate/Queue/Worker.php(375): Illuminate\Queue\Worker->process()
#39 /var/www/html/vendor/laravel/framework/src/Illuminate/Queue/Worker.php(173): Illuminate\Queue\Worker->runJob()
#40 /var/www/html/vendor/laravel/framework/src/Illuminate/Queue/Console/WorkCommand.php(150): Illuminate\Queue\Worker->daemon()
#41 /var/www/html/vendor/laravel/framework/src/Illuminate/Queue/Console/WorkCommand.php(134): Illuminate\Queue\Console\WorkCommand->runWorker()
#42 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(36): Illuminate\Queue\Console\WorkCommand->handle()
#43 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/Util.php(41): Illuminate\Container\BoundMethod::Illuminate\Container\{closure}()
#44 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(93): Illuminate\Container\Util::unwrapIfClosure()
#45 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(37): Illuminate\Container\BoundMethod::callBoundMethod()
#46 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/Container.php(651): Illuminate\Container\BoundMethod::call()
#47 /var/www/html/vendor/laravel/framework/src/Illuminate/Console/Command.php(139): Illuminate\Container\Container->call()
#48 /var/www/html/vendor/symfony/console/Command/Command.php(308): Illuminate\Console\Command->execute()
#49 /var/www/html/vendor/laravel/framework/src/Illuminate/Console/Command.php(124): Symfony\Component\Console\Command\Command->run()
#50 /var/www/html/vendor/symfony/console/Application.php(1002): Illuminate\Console\Command->run()
#51 /var/www/html/vendor/symfony/console/Application.php(299): Symfony\Component\Console\Application->doRunCommand()
#52 /var/www/html/vendor/symfony/console/Application.php(171): Symfony\Component\Console\Application->doRun()
#53 /var/www/html/vendor/laravel/framework/src/Illuminate/Console/Application.php(102): Symfony\Component\Console\Application->run()
#54 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php(129): Illuminate\Console\Application->run()
#55 /var/www/html/artisan(37): Illuminate\Foundation\Console\Kernel->handle()
#56 {main}
Next Aws\S3\Exception\S3Exception: Error executing "GetObject" on "https://s3.url.com/148/zHSTnZgnnXCnumAT.jpg"; AWS HTTP error: Client error: `GET https://s3.url.com/148/zHSTnZgnnXCnumAT.jpg` resulted in a `404 Not Found` response:
<?xml version="1.0" encoding="UTF-8"?><Error><Code>NoSuchKey</Code><Message>The specified key does not exist.</Message>< (truncated...)
NoSuchKey (client): The specified key does not exist. - <?xml version="1.0" encoding="UTF-8"?><Error><Code>NoSuchKey</Code><Message>The specified key does not exist.</Message></Error> in /var/www/html/vendor/aws/aws-sdk-php/src/WrappedHttpHandler.php:195
Stack trace:
#0 /var/www/html/vendor/aws/aws-sdk-php/src/WrappedHttpHandler.php(97): Aws\WrappedHttpHandler->parseError()
#1 /var/www/html/vendor/guzzlehttp/promises/src/Promise.php(204): Aws\WrappedHttpHandler->Aws\{closure}()
#2 /var/www/html/vendor/guzzlehttp/promises/src/Promise.php(169): GuzzleHttp\Promise\Promise::callHandler()
#3 /var/www/html/vendor/guzzlehttp/promises/src/RejectedPromise.php(42): GuzzleHttp\Promise\Promise::GuzzleHttp\Promise\{closure}()
#4 /var/www/html/vendor/guzzlehttp/promises/src/TaskQueue.php(48): GuzzleHttp\Promise\RejectedPromise::GuzzleHttp\Promise\{closure}()
#5 /var/www/html/vendor/guzzlehttp/guzzle/src/Handler/CurlMultiHandler.php(159): GuzzleHttp\Promise\TaskQueue->run()
#6 /var/www/html/vendor/guzzlehttp/guzzle/src/Handler/CurlMultiHandler.php(184): GuzzleHttp\Handler\CurlMultiHandler->tick()
#7 /var/www/html/vendor/guzzlehttp/promises/src/Promise.php(248): GuzzleHttp\Handler\CurlMultiHandler->execute()
#8 /var/www/html/vendor/guzzlehttp/promises/src/Promise.php(224): GuzzleHttp\Promise\Promise->invokeWaitFn()
#9 /var/www/html/vendor/guzzlehttp/promises/src/Promise.php(269): GuzzleHttp\Promise\Promise->waitIfPending()
#10 /var/www/html/vendor/guzzlehttp/promises/src/Promise.php(226): GuzzleHttp\Promise\Promise->invokeWaitList()
#11 /var/www/html/vendor/guzzlehttp/promises/src/Promise.php(269): GuzzleHttp\Promise\Promise->waitIfPending()
#12 /var/www/html/vendor/guzzlehttp/promises/src/Promise.php(226): GuzzleHttp\Promise\Promise->invokeWaitList()
#13 /var/www/html/vendor/guzzlehttp/promises/src/Promise.php(62): GuzzleHttp\Promise\Promise->waitIfPending()
#14 /var/www/html/vendor/aws/aws-sdk-php/src/AwsClientTrait.php(58): GuzzleHttp\Promise\Promise->wait()
#15 /var/www/html/vendor/league/flysystem-aws-s3-v3/AwsS3V3Adapter.php(499): Aws\AwsClient->execute()
#16 /var/www/html/vendor/league/flysystem-aws-s3-v3/AwsS3V3Adapter.php(262): League\Flysystem\AwsS3V3\AwsS3V3Adapter->readObject()
#17 /var/www/html/vendor/league/flysystem/src/Filesystem.php(79): League\Flysystem\AwsS3V3\AwsS3V3Adapter->readStream()
#18 /var/www/html/vendor/laravel/framework/src/Illuminate/Filesystem/FilesystemAdapter.php(588): League\Flysystem\Filesystem->readStream()
#19 /var/www/html/vendor/spatie/laravel-medialibrary/src/MediaCollections/Filesystem.php(180): Illuminate\Filesystem\FilesystemAdapter->readStream()
#20 /var/www/html/vendor/spatie/laravel-medialibrary/src/MediaCollections/Filesystem.php(185): Spatie\MediaLibrary\MediaCollections\Filesystem->getStream()
#21 /var/www/html/vendor/spatie/laravel-medialibrary/src/Conversions/FileManipulator.php(57): Spatie\MediaLibrary\MediaCollections\Filesystem->copyFromMediaLibrary()
#22 /var/www/html/vendor/spatie/laravel-medialibrary/src/Conversions/Jobs/PerformConversionsJob.php(27): Spatie\MediaLibrary\Conversions\FileManipulator->performConversions()
#23 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(36): Spatie\MediaLibrary\Conversions\Jobs\PerformConversionsJob->handle()
#24 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/Util.php(41): Illuminate\Container\BoundMethod::Illuminate\Container\{closure}()
#25 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(93): Illuminate\Container\Util::unwrapIfClosure()
#26 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(37): Illuminate\Container\BoundMethod::callBoundMethod()
#27 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/Container.php(651): Illuminate\Container\BoundMethod::call()
#28 /var/www/html/vendor/laravel/framework/src/Illuminate/Bus/Dispatcher.php(128): Illuminate\Container\Container->call()
#29 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(141): Illuminate\Bus\Dispatcher->Illuminate\Bus\{closure}()
#30 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(116): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}()
#31 /var/www/html/vendor/laravel/framework/src/Illuminate/Bus/Dispatcher.php(132): Illuminate\Pipeline\Pipeline->then()
#32 /var/www/html/vendor/laravel/framework/src/Illuminate/Queue/CallQueuedHandler.php(124): Illuminate\Bus\Dispatcher->dispatchNow()
#33 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(141): Illuminate\Queue\CallQueuedHandler->Illuminate\Queue\{closure}()
#34 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(116): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}()
#35 /var/www/html/vendor/laravel/framework/src/Illuminate/Queue/CallQueuedHandler.php(126): Illuminate\Pipeline\Pipeline->then()
#36 /var/www/html/vendor/laravel/framework/src/Illuminate/Queue/CallQueuedHandler.php(70): Illuminate\Queue\CallQueuedHandler->dispatchThroughMiddleware()
#37 /var/www/html/vendor/laravel/framework/src/Illuminate/Queue/Jobs/Job.php(98): Illuminate\Queue\CallQueuedHandler->call()
#38 /var/www/html/vendor/laravel/framework/src/Illuminate/Queue/Worker.php(425): Illuminate\Queue\Jobs\Job->fire()
#39 /var/www/html/vendor/laravel/framework/src/Illuminate/Queue/Worker.php(375): Illuminate\Queue\Worker->process()
#40 /var/www/html/vendor/laravel/framework/src/Illuminate/Queue/Worker.php(173): Illuminate\Queue\Worker->runJob()
#41 /var/www/html/vendor/laravel/framework/src/Illuminate/Queue/Console/WorkCommand.php(150): Illuminate\Queue\Worker->daemon()
#42 /var/www/html/vendor/laravel/framework/src/Illuminate/Queue/Console/WorkCommand.php(134): Illuminate\Queue\Console\WorkCommand->runWorker()
#43 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(36): Illuminate\Queue\Console\WorkCommand->handle()
#44 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/Util.php(41): Illuminate\Container\BoundMethod::Illuminate\Container\{closure}()
#45 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(93): Illuminate\Container\Util::unwrapIfClosure()
#46 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(37): Illuminate\Container\BoundMethod::callBoundMethod()
#47 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/Container.php(651): Illuminate\Container\BoundMethod::call()
#48 /var/www/html/vendor/laravel/framework/src/Illuminate/Console/Command.php(139): Illuminate\Container\Container->call()
#49 /var/www/html/vendor/symfony/console/Command/Command.php(308): Illuminate\Console\Command->execute()
#50 /var/www/html/vendor/laravel/framework/src/Illuminate/Console/Command.php(124): Symfony\Component\Console\Command\Command->run()
#51 /var/www/html/vendor/symfony/console/Application.php(1002): Illuminate\Console\Command->run()
#52 /var/www/html/vendor/symfony/console/Application.php(299): Symfony\Component\Console\Application->doRunCommand()
#53 /var/www/html/vendor/symfony/console/Application.php(171): Symfony\Component\Console\Application->doRun()
#54 /var/www/html/vendor/laravel/framework/src/Illuminate/Console/Application.php(102): Symfony\Component\Console\Application->run()
#55 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php(129): Illuminate\Console\Application->run()
#56 /var/www/html/artisan(37): Illuminate\Foundation\Console\Kernel->handle()
#57 {main}
Next League\Flysystem\UnableToReadFile: Unable to read file from location: 148/zHSTnZgnnXCnumAT.jpg. in /var/www/html/vendor/league/flysystem/src/UnableToReadFile.php:24
Stack trace:
#0 /var/www/html/vendor/league/flysystem-aws-s3-v3/AwsS3V3Adapter.php(501): League\Flysystem\UnableToReadFile::fromLocation()
#1 /var/www/html/vendor/league/flysystem-aws-s3-v3/AwsS3V3Adapter.php(262): League\Flysystem\AwsS3V3\AwsS3V3Adapter->readObject()
#2 /var/www/html/vendor/league/flysystem/src/Filesystem.php(79): League\Flysystem\AwsS3V3\AwsS3V3Adapter->readStream()
#3 /var/www/html/vendor/laravel/framework/src/Illuminate/Filesystem/FilesystemAdapter.php(588): League\Flysystem\Filesystem->readStream()
#4 /var/www/html/vendor/spatie/laravel-medialibrary/src/MediaCollections/Filesystem.php(180): Illuminate\Filesystem\FilesystemAdapter->readStream()
#5 /var/www/html/vendor/spatie/laravel-medialibrary/src/MediaCollections/Filesystem.php(185): Spatie\MediaLibrary\MediaCollections\Filesystem->getStream()
#6 /var/www/html/vendor/spatie/laravel-medialibrary/src/Conversions/FileManipulator.php(57): Spatie\MediaLibrary\MediaCollections\Filesystem->copyFromMediaLibrary()
#7 /var/www/html/vendor/spatie/laravel-medialibrary/src/Conversions/Jobs/PerformConversionsJob.php(27): Spatie\MediaLibrary\Conversions\FileManipulator->performConversions()
#8 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(36): Spatie\MediaLibrary\Conversions\Jobs\PerformConversionsJob->handle()
#9 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/Util.php(41): Illuminate\Container\BoundMethod::Illuminate\Container\{closure}()
#10 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(93): Illuminate\Container\Util::unwrapIfClosure()
#11 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(37): Illuminate\Container\BoundMethod::callBoundMethod()
#12 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/Container.php(651): Illuminate\Container\BoundMethod::call()
#13 /var/www/html/vendor/laravel/framework/src/Illuminate/Bus/Dispatcher.php(128): Illuminate\Container\Container->call()
#14 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(141): Illuminate\Bus\Dispatcher->Illuminate\Bus\{closure}()
#15 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(116): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}()
#16 /var/www/html/vendor/laravel/framework/src/Illuminate/Bus/Dispatcher.php(132): Illuminate\Pipeline\Pipeline->then()
#17 /var/www/html/vendor/laravel/framework/src/Illuminate/Queue/CallQueuedHandler.php(124): Illuminate\Bus\Dispatcher->dispatchNow()
#18 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(141): Illuminate\Queue\CallQueuedHandler->Illuminate\Queue\{closure}()
#19 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(116): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}()
#20 /var/www/html/vendor/laravel/framework/src/Illuminate/Queue/CallQueuedHandler.php(126): Illuminate\Pipeline\Pipeline->then()
#21 /var/www/html/vendor/laravel/framework/src/Illuminate/Queue/CallQueuedHandler.php(70): Illuminate\Queue\CallQueuedHandler->dispatchThroughMiddleware()
#22 /var/www/html/vendor/laravel/framework/src/Illuminate/Queue/Jobs/Job.php(98): Illuminate\Queue\CallQueuedHandler->call()
#23 /var/www/html/vendor/laravel/framework/src/Illuminate/Queue/Worker.php(425): Illuminate\Queue\Jobs\Job->fire()
#24 /var/www/html/vendor/laravel/framework/src/Illuminate/Queue/Worker.php(375): Illuminate\Queue\Worker->process()
#25 /var/www/html/vendor/laravel/framework/src/Illuminate/Queue/Worker.php(173): Illuminate\Queue\Worker->runJob()
#26 /var/www/html/vendor/laravel/framework/src/Illuminate/Queue/Console/WorkCommand.php(150): Illuminate\Queue\Worker->daemon()
#27 /var/www/html/vendor/laravel/framework/src/Illuminate/Queue/Console/WorkCommand.php(134): Illuminate\Queue\Console\WorkCommand->runWorker()
#28 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(36): Illuminate\Queue\Console\WorkCommand->handle()
#29 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/Util.php(41): Illuminate\Container\BoundMethod::Illuminate\Container\{closure}()
#30 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(93): Illuminate\Container\Util::unwrapIfClosure()
#31 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(37): Illuminate\Container\BoundMethod::callBoundMethod()
#32 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/Container.php(651): Illuminate\Container\BoundMethod::call()
#33 /var/www/html/vendor/laravel/framework/src/Illuminate/Console/Command.php(139): Illuminate\Container\Container->call()
#34 /var/www/html/vendor/symfony/console/Command/Command.php(308): Illuminate\Console\Command->execute()
#35 /var/www/html/vendor/laravel/framework/src/Illuminate/Console/Command.php(124): Symfony\Component\Console\Command\Command->run()
#36 /var/www/html/vendor/symfony/console/Application.php(1002): Illuminate\Console\Command->run()
#37 /var/www/html/vendor/symfony/console/Application.php(299): Symfony\Component\Console\Application->doRunCommand()
#38 /var/www/html/vendor/symfony/console/Application.php(171): Symfony\Component\Console\Application->doRun()
#39 /var/www/html/vendor/laravel/framework/src/Illuminate/Console/Application.php(102): Symfony\Component\Console\Application->run()
#40 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php(129): Illuminate\Console\Application->run()
#41 /var/www/html/artisan(37): Illuminate\Foundation\Console\Kernel->handle()
#42 {main}
```

Edit: Clearing cache and restarting workers seems to have fixed it.

**Comment by SlothLady** [comment]
```php
<?php

namespace App\MediaLibrary;

use Spatie\MediaLibrary\MediaCollections\Models\Media;
use Spatie\MediaLibrary\Support\PathGenerator\DefaultPathGenerator;

class CustomPathGenerator extends DefaultPathGenerator
{
    protected function getBasePath(Media $media): string
    {
        $prefix = config('media-library.prefix', '');

        if ($prefix !== '') {
            return $prefix.'/'.$media->getAttribute('uuid');
        }

        return $media->uuid;
    }
}
```

This works in the latest version of spatie media library however I'm finding that using uuids is horribly inefficient, like 7x higher application time inefficient and I can't figure out why.

**Reply by SlothLady** [reply]
Nvm it works fine, no performance hit. It was caused by something else
